### DISCUSSION TITLE
When will QUIC protocol be supported on all platforms?

### DISCUSSION THREAD

**Discussion by ShreyasZare** [dotnet/runtime#79736] [Q&A]
.NET 7 has added support for QUIC protocol which right now works only on Windows 11 and Windows Server 2022 out of the box. I have read that it is supported on Linux but does not work and any dependency that is required to be installed is also not documented. It also seems to be still in preview as it requires the code to have `[RequiresPreviewFeatures]` attribute set to compile.

So, the question is when can we expect QUIC support to be generally available on all supported platforms?

**Comment by wfurt** [comment]
https://devblogs.microsoft.com/dotnet/dotnet-7-networking-improvements/

Perhaps we should document the native dependencies in API docs @ManickaP ?

Note that `supported` and `RequiresPreviewFeatures` are two different things. We will support Quic with .NET 7. 
The `RequiresPreviewFeatures` is hint that we still may do some API adjustments. 

Now , "All" is overloaded question. There is currently no plan for mobile platforms and from .NET prospective the support depends on MsQuic availability. WASM is also problematic AFAIK. Older Windows are problematic. There is https://github.com/dotnet/runtime/issues/69978 but it will not work out of the box e.g. you would need to work on packaging and dependencies. macOS should work but there are no binary packages, primarily because signing and packaging issues.

**Reply by Fabi** [reply]
Sadly stuff like MsQuic is used in .NET. That's generally the wrong way to handle things in .NET. Those things should be implemented 100% in C#. Same goes for cryptography like AES etc. But the mistake is made over and over again and relying on OS depend implementations or libraries. That's the worst they do here

**Reply by wfurt** [reply]
It is LOT of work and comes with many caveats. Since .NET is open source project, you can start contributing.

**Reply by davidfowl** [reply]
Generally, I agree with this sentiment @Fabi but as @wfurt said, there are lots of tradeoffs.

**Comment by ShreyasZare** [comment]
Thanks for the response. I am working to implement DNS-over-QUIC support for my projects so hoping that QUIC support improves in .NET.

I did try installing `libmsquic` on Ubuntu Server 22.04 and it did not work so not sure if there is any thing else that needs to be installed or configured. It would be best if there can be some documentation available on installing the dependencies to make it work on Linux for both x86-64 and arm7 platforms. 

I understand that adding support for all platforms may not be feasible but at least popular ones should be supported. Like right now Windows 10 which has a huge install base is not supported.

**Reply by wfurt** [reply]
msquic needs OpenSSL 1.1.1 and it was not available as dependency on Ubuntu22 last time I check. You can contact the distro vendor to provide compat version like other distributions, switch version/distro or wrestle it in ways like other people do (just search internet)

**Reply by ShreyasZare** [reply]
> msquic needs OpenSSL 1.1.1

Thanks for the info. This dependency does not seems to be mentioned in the msquic repo. It would be nice to get it listed in the readme.

Now with the OpenSSL dependency in place, QUIC should work on Linux. So, the only popular platform that remains is Windows 10. Is there support for Windows 10 planned in later .NET 7 updates or .NET 8, or is it unclear at the moment?

**Reply by ShreyasZare** [reply]
Just confirmed on my updated Ubuntu 20.04 test VM and it shows:
`OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)`

I have `libmsquic` installed via Microsoft package feed.

Still getting the error:
```
System.PlatformNotSupportedException: System.Net.Quic is not supported on this platform.
   at System.Net.Quic.QuicConnection.ConnectAsync(QuicClientConnectionOptions options, CancellationToken cancellationToken)
```

So, is there any other dependency or config required apart from `libmsquic`?

**Reply by MichalPetryka** [reply]
@ShreyasZare `OpenSSL 1.1.1`, not `OpenSSL 3.0`.

**Reply by Fabi** [reply]
> @ShreyasZare `OpenSSL 1.1.1`, not `OpenSSL 3.0`.

that gives us the issue that most recent distros all ship with openssl 3 and u have to build openssl 1 yourself. Also I wonder if that limitation is required at all. Most things are 1:1 compatible in both versions

**Reply by MichalPetryka** [reply]
> > @ShreyasZare `OpenSSL 1.1.1`, not `OpenSSL 3.0`.
> 
> that gives us the issue that most recent distros all ship with openssl 3 and u have to build openssl 1 yourself. Also I wonder if that limitation is required at all. Most things are 1:1 compatible in both versions

They absolutely are not compatible, OpenSSL 3.0 literally shuffled exports between the 2 `.so` files.

**Reply by Fabi** [reply]
> > > @ShreyasZare `OpenSSL 1.1.1`, not `OpenSSL 3.0`.
> > 
> > 
> > that gives us the issue that most recent distros all ship with openssl 3 and u have to build openssl 1 yourself. Also I wonder if that limitation is required at all. Most things are 1:1 compatible in both versions
> 
> They absolutely are not compatible, OpenSSL 3.0 literally shuffled exports between the 2 `.so` files.

But that doesn't matter at all when you import by name.

**Reply by wfurt** [reply]
AFAIk support for OpenSSL 3.0 is coming to msquic but it is not production ready. Back in a day when everybody were switching from OpenSSL 1.0 to 1.1 distributions had both versions available. This is pretty simple. It seems like Ubuntu is only one that did not provide combat package this time. More discussion here: https://github.com/dotnet/arcade/issues/10253 You can look at https://askubuntu.com/questions/1403837/how-do-i-use-openssl-1-1-1-in-ubuntu-22-04 if you do not want to make your compilation.
Eventually the library should be provided just like another OS package IMHO. But since Quic is in early adopter state that did not happen yet. You can open issue for you favorite distribution to make it happen.

**Comment by ShreyasZare** [comment]
Based on the discussions here, I have written a [blog post guide](https://blog.technitium.com/2023/02/configuring-dns-over-quic-and-https3.html) for my .NET based project that helps users to enable QUIC support which someone here may find useful.

**Comment by wfurt** [comment]
MsQuic 2.1.8 was released on Friday and the Linux package does have proper dependencies e.g. it won’t silently install but not function if there is no OpenSSL 1.1 (e.g. Ubuntu 22.04)
The instructions from blog would no longer work if followed exactly @ShreyasZare 
 
```shell
$sudo apt-get install libmsquic
Some packages could not be installed. This may mean that you have
 
The following packages have unmet dependencies:
libmsquic : Depends: libssl1.1 but it is not installable
E: Unable to correct problems, you have held broken packages.
 ```

You can either force install without dependencies or swap the order. (see example below)

```bash 
$sudo dpkg -i  libssl1.1_1.1.1f-1ubuntu2.17_amd64.deb
Preparing to unpack libssl1.1_1.1.1f-1ubuntu2.17_amd64.deb ...
Unpacking libssl1.1:amd64 (1.1.1f-1ubuntu2.17) ...
Setting up libssl1.1:amd64 (1.1.1f-1ubuntu2.17) ...
Processing triggers for libc-bin (2.35-0ubuntu3) ...
 
$ sudo apt-get install libmsquic
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages will be upgraded:
  libmsquic
1 upgraded, 0 newly installed, 0 to remove and 224 not upgraded.
Need to get 4,127 kB of archives.
After this operation, 6,144 B of additional disk space will be used.
Get:1 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main amd64 libmsquic amd64 2.1.8 [4,127 kB]
Fetched 4,127 kB in 1s (3,039 kB/s)
(Reading database ... 269943 files and directories currently installed.)
Preparing to unpack .../libmsquic_2.1.8_amd64.deb ...
Unpacking libmsquic (2.1.8) over (2.1.7) ...
Setting up libmsquic (2.1.8) ...
 ```

Support for OpenSSL 3.x should be in MsQuic 2.2 (works in current main branch)

**Reply by ShreyasZare** [reply]
Thanks for letting me know!
